### PR TITLE
feature (auth/ldap): add repo and reading an auth method

### PR DESCRIPTION
**Pull request by jimlambrt**

### CODE REVIEW

File: internal/auth/ldap/repository_auth_method_read.go [hunk 1]
```diff
@@ -0,0 +1,235 @@
...
+	"github.com/hashicorp/go-kms-wrapping/v2/extras/structwrapping"
+)
+
+// LookupAuthMethod will lookup an auth method in the repo, along with its
+// associated Value Objects of SigningAlgs, CallbackUrls, AudClaims and
+// Certificates. If it's not found, it will return nil, nil.  The
+// WithUnauthenticatedUser options is supported and all other options are
+// ignored.
+func (r *Repository) LookupAuthMethod(ctx context.Context, publicId string, opt ...Option) (*AuthMethod, error) {
+	const op = "oidc.(Repository).LookupAuthMethod"
```
Comment (irenarindos) [inline internal/auth/ldap/repository_auth_method_read.go:21]:
```suggestion
	const op = "ldap.(Repository).LookupAuthMethod"
```

File: internal/auth/ldap/repository_auth_method_read.go [hunk 2]
```diff
@@ -0,0 +1,235 @@
...
+		return nil, errors.Wrap(ctx, err, op)
+	}
+	return r.lookupAuthMethod(ctx, publicId, WithUnauthenticatedUser(opts.withUnauthenticatedUser))
+}
+
+// ListAuthMethods returns a slice of AuthMethods for the scopeId. The
+// WithUnauthenticatedUser, WithLimit and WithOrder options are supported and
+// all other options are ignored.
+func (r *Repository) ListAuthMethods(ctx context.Context, scopeIds []string, opt ...Option) ([]*AuthMethod, error) {
+	const op = "oidc.(Repository).ListAuthMethods"
```
Comment (irenarindos) [inline internal/auth/ldap/repository_auth_method_read.go:36]:
```suggestion
	const op = "ldap.(Repository).ListAuthMethods"
```

File: internal/auth/ldap/repository_auth_method_read.go [hunk 3]
```diff
@@ -0,0 +1,235 @@
...
+	authMethods, err := r.getAuthMethods(ctx, "", scopeIds, opt...)
+	if err != nil {
+		return nil, errors.Wrap(ctx, err, op)
+	}
+	return authMethods, nil
+}
+
+// lookupAuthMethod will lookup a single auth method
+func (r *Repository) lookupAuthMethod(ctx context.Context, authMethodId string, opt ...Option) (*AuthMethod, error) {
+	const op = "oidc.(Repository).lookupAuthMethod"
```
Comment (irenarindos) [inline internal/auth/ldap/repository_auth_method_read.go:49]:
```suggestion
	const op = "ldap.(Repository).lookupAuthMethod"
```

File: internal/auth/ldap/repository_auth_method_read.go [hunk 4]
```diff
@@ -0,0 +1,235 @@
...
+// WithLimit and WithOrder options are supported and all other options are
+// ignored.
+//
+// The AuthMethod returned has its value objects populated (SigningAlgs,
+// CallbackUrls, AudClaims and Certificates).  The AuthMethod returned has its
+// IsPrimaryAuthMethod bool set.
+//
+// When no record is found it returns nil, nil
+func (r *Repository) getAuthMethods(ctx context.Context, authMethodId string, scopeIds []string, opt ...Option) ([]*AuthMethod, error) {
+	const op = "oidc.(Repository).getAuthMethods"
```
Comment (irenarindos) [inline internal/auth/ldap/repository_auth_method_read.go:77]:
```suggestion
	const op = "ldap.(Repository).getAuthMethods"
```

File: internal/auth/ldap/repository_auth_method_read.go [hunk 5]
```diff
@@ -0,0 +1,235 @@
...
+		am.BindPasswordHmac = agg.BindPasswordHmac
+
+		authMethods = append(authMethods, &am)
+	}
+	return authMethods, nil
+}
+
+// authMethodAgg is a view that aggregates the auth method's value objects if
+// the value object can have multiple values like Urls and Certs, then the
+// string field is delimited with the aggregateDelimiter of "|"
```
Comment (irenarindos) [inline internal/auth/ldap/repository_auth_method_read.go:200]:
```suggestion
// authMethodAgg is a view that aggregates the auth method's value objects.
// If the value object can have multiple values like Urls and Certs, then the
// string field is delimited with the aggregateDelimiter of "|"
```

**Review by irenarindos — CHANGES_REQUESTED** [review]

File: internal/auth/ldap/repository_auth_method_read.go (hunk 1 above)
Comment (jimlambrt) [reply internal/auth/ldap/repository_auth_method_read.go:21]: ty!

File: internal/auth/ldap/repository_auth_method_read.go (hunk 2 above)
Comment (jimlambrt) [reply internal/auth/ldap/repository_auth_method_read.go:36]: ty!

File: internal/auth/ldap/repository_auth_method_read.go (hunk 3 above)
Comment (jimlambrt) [reply internal/auth/ldap/repository_auth_method_read.go:49]: ty!

File: internal/auth/ldap/repository_auth_method_read.go (hunk 4 above)
Comment (jimlambrt) [reply internal/auth/ldap/repository_auth_method_read.go:77]: ty!

File: internal/auth/ldap/repository_auth_method_read.go (hunk 5 above)
Comment (jimlambrt) [reply internal/auth/ldap/repository_auth_method_read.go]: ty!

File: internal/auth/ldap/repository_auth_method_read.go [hunk 6]
```diff
@@ -0,0 +1,235 @@
...
+	var where []string
+	switch {
+	case authMethodId != "":
+		where, args = append(where, "public_id = ?"), append(args, authMethodId)
+	default:
+		where, args = append(where, "scope_id in(?)"), append(args, scopeIds)
+	}
+
+	if opts.withUnauthenticatedUser {
+		where, args = append(where, "state = ?"), append(args, string(ActivePublicState))
```
Comment (johanbrandhorst) [inline internal/auth/ldap/repository_auth_method_read.go:117]: Why does this imply this `where` statement? Could you add a small comment here?

Comment (jimlambrt) [reply internal/auth/ldap/repository_auth_method_read.go:117]: I'll add a comment.

File: internal/auth/ldap/repository_auth_method_read.go [hunk 7]
```diff
@@ -0,0 +1,235 @@
...
+	"github.com/hashicorp/boundary/internal/db"
+	"github.com/hashicorp/boundary/internal/db/timestamp"
+	"github.com/hashicorp/boundary/internal/errors"
+	"github.com/hashicorp/boundary/internal/kms"
+	"github.com/hashicorp/go-kms-wrapping/v2/extras/structwrapping"
+)
+
+// LookupAuthMethod will lookup an auth method in the repo, along with its
+// associated Value Objects of SigningAlgs, CallbackUrls, AudClaims and
+// Certificates. If it's not found, it will return nil, nil.  The
```
[the remark this answers is not in the document]
Comment (jimlambrt) [reply internal/auth/ldap/repository_auth_method_read.go:17]: While I understand your consternation. I don't think we should change this pattern in just one place... making it inconsistent with the rest of the code base.  Also, I'm happy to engage in a larger discussion about changing this pattern and refactoring the code base but that discussion needs to include the entire team.

[the remark this answers is not in the document]
Comment (johanbrandhorst) [reply internal/auth/ldap/repository_auth_method_read.go:17]: Sounds good, I've pinged an internal channel to kick off this discussion. Would be great if could change it everywhere :).

**Review by johanbrandhorst — APPROVED** [review]

File: internal/db/schema/migrations/oss/postgres/60/01_ldap.up.sql [hunk 8]
```diff
@@ -368,4 +368,65 @@ values
...
+  am.discover_dn,
+  am.anon_group_search,
+  am.upn_domain,
+  -- the string_agg(..) column will be null if there are no associated value objects
+  string_agg(distinct url.url, '|') as urls,
+  string_agg(distinct cert.certificate, '|') as certs,
+
+  -- the rest of the fields are zero to one relationships that are stored in
+  -- related tables.
+  string_agg(distinct uc.user_dn, '|') as user_dn,
```
Comment (irenarindos) [inline internal/db/schema/migrations/oss/postgres/60/01_ldap.up.sql:405]: I thought only URLs and certs could have multiple delimited values?

**Review by irenarindos — APPROVED** [review]

**Pull request merged**